### ISSUE TITLE
redundant-closure-for-method-calls suggests <dyn trait>::fn which does not compile

### ISSUE THREAD

**Issue by matthiaskrgr** [rust-lang/rust-clippy#10446]
### Summary

.

### Lint Name

redundant-closure-for-method-calls 

### Reproducer

I tried this code:

```rust
#![feature(error_generic_member_access, provide_any)]

use std::backtrace::{Backtrace};
use std::error::Error;

const _: fn(&dyn Error) -> Option<&Backtrace> = |err| err.request_ref::<Backtrace>();

pub fn main() {}

```

I saw this happen:

```
warning: redundant closure
 --> src/main.rs:6:49
  |
6 | const _: fn(&dyn Error) -> Option<&Backtrace> = |err| err.request_ref::<Backtrace>();
  |                                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: replace the closure with the method itself: `<dyn std::error::Error>::request_ref`
  |
```

Suggested code does not compile
````
The following errors were reported:
error[E0308]: mismatched types
 --> src/main.rs:6:49
  |
6 | const _: fn(&dyn Error) -> Option<&Backtrace> = <dyn std::error::Error>::request_ref;
  |                                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ one type is more general than the other
  |
  = note: expected fn pointer `for<'a> fn(&'a (dyn std::error::Error + 'a)) -> std::option::Option<&'a std::backtrace::Backtrace>`
                found fn item `fn(&dyn std::error::Error) -> std::option::Option<&_> {<dyn std::error::Error>::request_ref::<_>}`
````

### Version

```text
rustc 1.69.0-nightly (b1719530f 2023-03-04)
binary: rustc
commit-hash: b1719530f44e3c8ec903f76020a52bd8764d5d10
commit-date: 2023-03-04
host: x86_64-unknown-linux-gnu
release: 1.69.0-nightly
LLVM version: 15.0.7
```
